### ISSUE TITLE
Interrupt long running queries

### ISSUE THREAD

**Issue by gennaios** [groue/GRDB.swift#703]
### What did you do?

Using live search fields with GRDB on iOS or macOS such as a NSSearchField can cause long running queries to continue running that are no longer relevant if the search field text has changed. It would be nice to interrupt only a specific query rather than all that occurs with using sqlite3_interrupt. There are Query Progress Callbacks using sqlite3_progress_handler. The docs note one progress handler can be defined at a time and setting a new one will cancel the old. Perhaps it would be sufficient if it were possible to execute SQL with some parameter to use a progress handler, and then any subsequent query as such would cancel the pending one. This has the benefit of not cancelling all queries such as possible reads or writes on other threads as would occur with sqlite3_interrupt.

https://www.sqlite.org/c3ref/progress_handler.html

### Environment

**GRDB flavor(s): GRDB
**GRDB version: 4.9.0
**Installation method: CocoaPods
**Xcode version: 11.3.1
**Swift version: 5.1.x I believe
**Platform(s) running GRDB: macOS
**macOS version running Xcode: 10.15.3

**Comment by groue** [comment]
Hello @gennaios,

This is an interesting request.

Will you please keep on your investigations, and suggest the draft for a public API, as well as an implementation plan that makes sure this public API can be implemented, so that we can turn this feature request of yours into a reality?

Thanks in advance for your help

**Comment by groue** [comment]
On top of that, a more detailed explanation of problems created by sqlite3_interrupt, or the way sqlite3_interrupt is exposed by GRDB, would be appreciated as well.

This extra request is justified by the fact that there may be room for improvement around sqlite3_interrupt. We have to know if our energy should be put there, or in sqlite3_progress_handler, and why.

**Comment by gennaios** [comment]
Hello groue,

Concerning sqlite3_interrupt, from the documentation :

> The sqlite3_interrupt(D) call is in effect until all currently running SQL statements on database connection D complete. Any new SQL statements that are started after the sqlite3_interrupt() call and before the running statement count reaches zero are interrupted as if they had been running prior to the sqlite3_interrupt() call. New SQL statements that are started after the running statement count reaches zero are not effected by the sqlite3_interrupt().

It seems that when sqlite3_interrupt is called and is in operation until finished, that is when the running statement count becomes 0, any current statements as well as any new statements added even after sqlite3_interrupt has been called but before it has completed, are terminated. From what I can tell, that may be appropriate for simpler apps that don't use a `DatabasePool` but in threaded applications that might have multiple read and writes occurring, any running statements will be terminated. Maybe there is some elegant way to wait for completion of certain statements before invoking sqlite3_interrupt though I am not sure.

The benefit of sqlite3_progress_handler is that one can pass it to a specific exec statement which will then periodically invoke the callback as well as check if a new one has been set, upon which only the specific statement invoked with a callback will be terminated.

Such is what I've determined from my understanding of the docs. Perhaps the progress handler is a newer addition to SQLite. At least for me, I can't think of cases where I would want to terminate all running statements but being able to terminate one, I think that'd be very useful.

I might not be the best to ask for how such might look as my programming ability is fairly limited, but perhaps there could be a new Callback type, and then one could invoke statements such as :

`db.execute(sql: …, arguments: …, callback: …)`

From the docs, I'm not sure exactly what is `callback function X` that is invoked when a statement is ran with a callback. Perhaps there is the possibility to either within the callback function to do stuff such as other tasks such as update a UI to show that a particular statement is running, e.g., for progress hence the name, or to somehow use it such that a new callback is passed along with a subsequent statement to cancel an old one. I'm not exactly sure. I've tried looking at the user mailing list a bit and haven't found that much help so far. I'll continue to look and perhaps some examples can be found on GitHub.

**Comment by groue** [comment]
> From what I can tell, that may be appropriate for simpler apps that don't use a DatabasePool but in threaded applications that might have multiple read and writes occurring, any running statements will be terminated.

This is not due to sqlite3_interrupt(). This is due to the only interrupt() method provided by GRDB, which is DatabaseQueue/Pool.interrupt(). In case of database pools, it indeed interrupts all existing connections, and won't let you target a specific query.

To me, this is not a problem with sqlite3_interrupt. Instead, this is a limitation of the GRDB api. We can discuss this limitation, but I still don't understand why sqlite3_progress_handler is required.

> The benefit of sqlite3_progress_handler is that one can pass it to a specific exec statement which will then periodically invoke the callback as well as check if a new one has been set, upon which only the specific statement invoked with a callback will be terminated.

You have to be much more specific in what you envision, and again explain why it is superior to sqlite3_interrupt.

> `db.execute(sql: …, arguments: …, callback: …)`

I don't quite understand how this fits your initial request, which was interrupting a query spawned by an NSSearchField. And why do you only focus on execute(sql:), which performs INSERT and UPDATE statements? What about SELECT? And what about users who don't use raw SQL statements, and prefer using the query interface?

---

Did you try something which as below?

```swift
// Start job that you want to interrupt eventually:
weak var dbToInterrupt: Database?
try dbPool.read { db in
    dbToInterrupt = db
    defer { dbToInterrupt = nil }
    let results = /* some stuff from db */
}

// When interruption is needed:
if let dbToInterrupt = dbToInterrupt {
    sqlite3_interrupt(dbToInterrupt.sqliteConnection)
}
```

If you think this does not fit, or if it creates problems, will you be specific about the reasons why?

**Comment by gennaios** [comment]
groue,

Perhaps there is some lack of understanding for us both. Initially, I had not understood your hint that such is possible with sqlite3_interrupt. Being a novice, fairly unfamiliar too with the API, I had not thought of such. This is a world where what some call speech is lacking. I think of Blanchot, particularly Le livre à venir and L'entretien infini, or Le degré zéro de l'écriture by Barthes. Think of the charter of L'Académie française, the purpose set forth by Richelieu. Half a millennia later, has language itself, that of everyday life, become more profound ? Do the conditions of everyday life encourage depth in grammar, complexity in seeing relations of us or things ? If such were the case, perhaps grammar would have become more profound such as that of Old French, not counting Latin or Ancient Greek. It took some time to decrypt some of the things you had said. Too, you may have realized there was some understanding I lacked given my response.

As such, I did not understand until you provided some code. Perhaps too you were hoping for as mentioned a draft for an API for sqlite3_progress_handler. With mind perhaps scattered, I had perhaps sensed but did not see the work required to implement such across all queries in all forms. Indeed, perhaps such would require a fair amount of work. Perhaps nice and beneficial for others someday.

For now, perhaps sqlite3_interrupt is indeed sufficient. Trying to figure out how to implement such, I have something as following. Queries are typically done as struct extensions. I have multiple tables, some one to many, some many to many, a view, and a FTS5. As such, I have a TableWork struct and a ViewWork struct. Imagine like the example of Book and Author, yet I have fields such as editor, translator, etc., publisher external, and so forth, with joins and group_concat for the view to simplify the search table (editors, translators, etc. in one field). For the NSTableView, a view model trying to follow MVVM. As such, the view controller calls table view model which calls the function as struct extension which performs the query. In progress trying to get this working, and so far it looks something like this :

``
class TableViewModel: NSObject {

    var works: [ViewWork] = []
     weak var searchDatabase: Database?

     func searchWorks() -> [ViewWork]? {
          …
          (searchDatabase, works) = ViewWork.search(database: searchDatabase, query: …)
          …
     }


class ViewController: NSViewController {
    public func filterTable() {
        DispatchQueue.global(qos: .userInteractive).async {
            guard let newWorks = self.viewModel.searchWorks() else {
                return
            }

            DispatchQueue.main.sync(execute: { () -> Void in
                self.viewModel.works = newWorks
                self.tableView.reloadData()
            })
        }
    }
}

extension ViewWork {
    static func search(database: Database?, query: String, …) ->(Database?, [ViewWork]) {
   …
   weak var databaseOperation: Database?
        if let searchDatabase = database {
            print("cancelling previous query")
            sqlite3_interrupt(searchDatabase.sqliteConnection)
        }

        do {
            try dbPool.read { db in
                databaseOperation = db
                items = try ViewWork.fetchAll(db, sql: sql, arguments: [query])
            }
        } catch {
            print("Error …)
        }

        return (databaseOperation, works)
    }
}
``

That as, given what I understand, the cancellable `Database` exists within the view model, is passed to where it is tested, and then returned so there is a reference at the next query. It seems to work though sometimes the `catch` block gets called. A `DispatchQueue` is used in the view controller so the UI remains responsive during search. Still working on it. Perhaps you could provide some insight, if something is missing or there could be some more elegant way. Thank you.

**Comment by gennaios** [comment]
Concerning the catch block that sometimes gets called with the above code : SQLite error 9 with statement. `The SQLITE_INTERRUPT result code indicates that an operation was interrupted by the sqlite3_interrupt() interface.` Perhaps it is because what I've tried to implement differs a bit from your suggestion though that is what I had changed to get it to work. Still trying.

**Comment by groue** [comment]
I'll make a longer answer, but:

> Concerning the catch block that sometimes gets called with the above code : SQLite error 9 with statement. The SQLITE_INTERRUPT result code indicates that an operation was interrupted by the sqlite3_interrupt() interface.

This is to be expected. Check https://github.com/groue/GRDB.swift/blob/master/README.md#interrupt-a-database

**Comment by groue** [comment]
So, @gennaios.

Yes, I've been quite elusive, and I admit my purpose has not quite been achieved.

This elusiveness was intended as leaving you some room so that you could elaborate on your needs, learn more from both interruption methods provided by SQLite, compare them and understand their differences, learn about GRDB apis (SQL, records, query interface), why not learn a little bit from GRDB internals as well, and eventually provide a high-level description of the problem to solve, and a path to a solution after you have mixed and merged your understandings on SQLite and GRDB.

I was hoping my role would be limited at providing some hints towards a sound and well-defined plan around interruption, aimed at helping application developers in their every day job, but not at defining it myself.

I know, of course, that the task is not that easy. But I myself knew nothing about SQLite when I started this library. And I'm keen to let good surprises come in :-)

> For now, perhaps sqlite3_interrupt is indeed sufficient.

Considering my limited understanding of sqlite3_interrupt and sqlite3_progress_handler after a few Google searches, this was my temporary conclusion as well.

> In progress trying to get this working, and so far it looks something like this :

So you're interrupting the current read before starting a new one. All right.

Warning: when the read is done, you do not reset `databaseOperation` to nil, as in my sample code. This means that you may interrupt any other read performed by this database connection. This looks like a bug.

Besides, you also have never shown that letting previous searches run to completion was indeed harmful to your application, and that you do need to perform this interruption. Optimizations are cool, but rarely necessary. Database pools can perform concurrent reads, which means that several reads can be executed in parallel: a running previous search does not prevent a new search from being performed. So do you really notice some undesired behavior, even if you spare a few CPU cycles? Which one is it? Is it the sign of some unexpected read contention, which happens when many concurrent reads fight for an access to the database and have to wait? The default number of concurrent reads is 5, which is already a big number, considering that requests, even full text requests, are usually quite fast.

**Comment by groue** [comment]
> I think of Blanchot, particularly Le livre à venir and L'entretien infini, or Le degré zéro de l'écriture by Barthes. Think of the charter of L'Académie française, the purpose set forth by Richelieu.

I forgot to say thank you for the reading list :-) Let's hope for the best.

**Comment by gennaios** [comment]
Hello groue,

I had not understood the threading model of SQLite and that a `DatabasePool` creates a separate connection on each read and write. Perhaps from my turn of phrase, you might have deduced such. As such indeed I believe for my case `sqlite3_interrupt` is sufficient.

What I currently have is the following :

```
    static func search(query: String, …) throws -> [ViewWork] {
        var works: [ViewWork] = []
        
        // …

        // Cancel previous query if still in progress
        if let searchDatabaseConnection = searchDatabaseConnection {
            sqlite3_interrupt(searchDatabaseConnection.sqliteConnection)
        }

        do {
            try dbPool.read { db in
                searchDatabaseConnection = db
                defer {
                    searchDatabaseConnection = nil
                }
                works = try ViewWork.fetchAll(db, sql: sql, arguments: [query])
            }
        } catch let error as DatabaseError {
            if error.resultCode == .SQLITE_INTERRUPT {
                print("Previous search interrupted: \(query).")
            }
        }
        return works
    }
```

Since this is a struct extension, as far as I can tell, `searchDatabaseConnection` needs to be a global to be accessible on subsequent calls. The only thing I am unsure about is in your example, you interrupt the connection after the query. Since it is re-assigned to the new connection, given my understanding, I'm not sure how such would work.

As far as optimization, my database is not exceedingly large though as it is approaching 5 GB, a query can potentially return many results. I too have multiple background `OperationQueue`s that may be fetching or updating data, and occasionally running an online backup. Perhaps that may not have a substantial impact though who knows. Indeed as recommended elsewhere, measurements are needed to tell.

As for progress callbacks, perhaps there may be cases when someone may need periodic tasks run during a particular connection, perhaps more useful in seeming here unneeded instances where one may be running multiple queries on the same connection and using some other module or their own, though as you can tell, I have no need at the moment.

**Comment by groue** [comment]
I recommend a look at the [GRDB Concurrency Guide](https://github.com/groue/GRDB.swift/blob/master/README.md#concurrency).

A Database Pool maintains one writer connection, and a pool of read-only connections. The lifetime of those connections is only partially under the control of the application. What you can rely on is that `dbPool.read` gives you some reader connection. Until the read is complete, no other `dbPool.read` will get the same connection. After the read is complete, other `dbPool.read` may get the same connection.

I'm not sure my help is needed any longer on this issue. I understand you may still have questions, but I recommend those questions remain about GRDB, and not about the best way to architecture your application.
